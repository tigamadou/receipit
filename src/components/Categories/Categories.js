import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Axios from 'axios';
import { setRecipes } from '../../redux/actions/index';
import Recipe from '../Recipe/Recipe';

const CategoriesComponent = ({ category, datas, setRecipes }) => {
  useEffect(() => {
    Axios.get(`https://www.themealdb.com/api/json/v1/1/filter.php?c=${category.strCategory}`)
      .then((response) => {
        // handle success
        setRecipes(response.data.meals);
        console.log(response);
      })
      .catch((error) => {
        // handle error
        console.log(error);
      })
      .then(() => {
        // always executed
      });
  }, []);
  return (
    <div>
      {datas.map((data) => (
        <Recipe data={data} key={data.id} />
      ))}
    </div>
  );
};

CategoriesComponent.propTypes = {
  category: PropTypes.instanceOf(Object),
  datas: PropTypes.arrayOf(PropTypes.instanceOf(Object)),
  setRecipes: PropTypes.func,
};
CategoriesComponent.defaultProps = {
  category: null,
  datas: [],
  setRecipes: null,
};
const mapDispatchToProps = (dispatch) => ({
  setRecipes: (category, recipes) => dispatch(setRecipes(category, recipes)),
});

const getDatas = (category, datas) => datas.filter((data) => data.id === category.id);
const mapStateToProps = (state) => ({
  category: state.category,
  datas: getDatas(state.category, state.datas),
});

export default connect(mapStateToProps, mapDispatchToProps)(CategoriesComponent);
