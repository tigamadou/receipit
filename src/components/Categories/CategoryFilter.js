import React from 'react';
import PropTypes from 'prop-types';

const CategoryFilter = ({ filter, setFilter }) => (
  <div>
    <p>Filtrer by</p>
    <select onChange={(e) => setFilter(e.target.value)} value={filter}>
      <option>ASC</option>
      <option>DESC</option>
    </select>
  </div>
);

CategoryFilter.defaultProps = {
  setFilter: null,
  filter: 'ASC',
};

CategoryFilter.propTypes = {
  setFilter: PropTypes.func,
  filter: PropTypes.string,
};

export default CategoryFilter;
