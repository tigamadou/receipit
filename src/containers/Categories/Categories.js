import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { useParams } from 'react-router-dom';
import Axios from 'axios';
import { toast } from 'react-toastify';
import { SetRecipes, SetHeader } from '../../redux/actions/index';
import Recipe from '../../components/Recipe/Recipe';
import CategoryFilter from '../../components/CategoryFilter/CategoryFilter';

const getDatas = (category, datas) => datas.filter((data) => data.strCategory === category);

const CategoriesComponent = ({
  datas, SetRecipes, SetHeader,
}) => {
  const { categoryName } = useParams();
  const [recipes, SetRecipesLocal] = useState([]);
  const [filter, setFilter] = useState('');

  const filterRecipes = (datas) => datas.filter(
    (data) => data.strMeal.toLowerCase().indexOf(filter.toLowerCase()) > -1,
  );

  const getRecipes = () => {
    Axios.get(`https://www.themealdb.com/api/json/v1/1/filter.php?c=${categoryName}`)
      .then((response) => {
        SetRecipes(categoryName, response.data.meals);
        SetRecipesLocal(filterRecipes(response.data.meals));
      })
      .catch(() => {
        toast.error('Oups! Something went wrong. Please try again later.', {
          position: 'top-right',
          autoClose: 5000,
          hideProgressBar: false,
          closeOnClick: true,
          pauseOnHover: true,
          draggable: true,
          progress: undefined,
        });
      });
  };

  const init = async () => {
    const category = getDatas(categoryName, datas)[0];
    if (category && category.recipes) {
      SetRecipesLocal(filterRecipes(category.recipes));
      return;
    }
    getRecipes();
  };

  useEffect(() => {
    SetHeader({ title: `${categoryName}`, back: '/' });
    init();
  }, [filter]);

  return (
    <>
      <CategoryFilter setFilter={setFilter} filter={filter} />
      <div className="container" data-testid="categories">
        <div className="cards">
          {(recipes && recipes.length > 0)
            ? (
              recipes.map((data) => (
                <Recipe data={data} key={data.idMeal} />
              ))
            )
            : (
              <div className="card no_result">
                No result found!
              </div>
            )}
        </div>
      </div>
    </>
  );
};

CategoriesComponent.propTypes = {
  datas: PropTypes.arrayOf(PropTypes.instanceOf(Object)),
  SetRecipes: PropTypes.func,
  SetHeader: PropTypes.func,
};
CategoriesComponent.defaultProps = {
  datas: [],
  SetRecipes: null,
  SetHeader: null,
};
const mapDispatchToProps = (dispatch) => ({
  SetRecipes: (category, recipes) => dispatch(SetRecipes(category, recipes)),
  SetHeader: (header) => dispatch(SetHeader(header)),
});

const mapStateToProps = (state) => ({
  datas: state.datas,
});

export default connect(mapStateToProps, mapDispatchToProps)(CategoriesComponent);
