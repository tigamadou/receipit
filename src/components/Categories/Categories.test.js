import React from 'react';
import {
  MemoryRouter as Router,
  Route,
} from 'react-router-dom';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Categories from './Categories';
import storeReducer from '../../redux/store';

const store = storeReducer();

describe('Categories component tests', () => {
  it('renders Categories component ', () => {
    render(<Provider store={store}><Router><Route><Categories /></Route></Router></Provider>);
    // eslint-disable-next-line no-unused-expressions
    expect(screen.getByTestId('categories')).toBeInTheDocument;
  });

  test('should match with snapshot', () => {
    const tree = renderer
      .create(<Provider store={store}><Router><Route><Categories /></Route></Router></Provider>);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
