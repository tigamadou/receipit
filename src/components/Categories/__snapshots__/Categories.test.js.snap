// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories component tests should match with snapshot 1`] = `
Array [
  <div
    id="categorySearchFilter"
  >
    <input
      className="filter"
      onChange={[Function]}
      placeholder="Type the recipe name"
      value=""
    />
  </div>,
  <div
    className="container"
    data-testid="categories"
  >
    <div
      className="cards"
    />
  </div>,
]
`;
